core/src/wsv.rs

```diff
@@ -279,23 +279,155 @@ impl WorldStateView {
+
+                // Cloning and clearing `self.ids_write` so that `handle_` call won't deadlock
+                let mut matched_ids = Vec::new();
+                std::mem::swap(&mut matched_ids, &mut self.world.triggers.matched_ids);
+
+                for (event_type, id) in matched_ids {
+                    // Ignoring `None` variant cause this means that action was deleted after `handle_`
+                    // call and before `inspect_matching()` call
+                    let _ = match event_type {
+                        Event::Data(_) => {
+                            let dangerous_reference = unsafe {
+                                let ptr = &self.world.triggers as *const TriggerSet; // TODO: REALLY FIX THIS
+                                &*ptr
+                            };
+
+                            dangerous_reference.data_triggers.get(&id).map(|entry| {
+                                let action = entry.value();
+                                if let Repeats::Exactly(atomic) = action.repeats() {
+                                    if atomic.get() == 0 {
+                                        return;
+                                    }
+                                }
+
+                                match self.process_trigger(action, event_type) {
+                                    Ok(()) => succeed.push(id),
+                                    Err(err) => errors.push(err),
+                                }
+                            })
+                        }
+                        Event::Pipeline(_) => {
+                            let dangerous_reference = unsafe {
+                                let ptr = &self.world.triggers as *const TriggerSet; // TODO: REALLY FIX THIS
+                                &*ptr
+                            };
+
+                            dangerous_reference.pipeline_triggers.get(&id).map(|entry| {
+                                let action = entry.value();
+                                if let Repeats::Exactly(atomic) = action.repeats() {
+                                    if atomic.get() == 0 {
+                                        return;
+                                    }
+                                }
+
+                                match self.process_trigger(action, event_type) {
+                                    Ok(()) => succeed.push(id),
+                                    Err(err) => errors.push(err),
+                                }
+                            })
+                        }
+                        Event::Time(_) => {
+                            let dangerous_reference = unsafe {
+                                let ptr = &self.world.triggers as *const TriggerSet; // TODO: REALLY FIX THIS
+                                &*ptr
+                            };
+
+                            dangerous_reference.time_triggers.get(&id).map(|entry| {
+                                let action = entry.value();
```

Comment:
Each of these is repeated. You can compress this with either a generic or a macro.

Reply:
Not worth it for code that it transitory. Also in my experience this type of pattern tends to diverge and so you'll most likely have to remove the macro later. It's not like you have to write the code more than once so the savings by a macro are not large.

Reply:
Resolved with rollback.